### PR TITLE
Fix/init mint amount

### PR DESCRIPTION
**Pull request by njerschow**
## 1. Overview

This PR includes

- initial mint amount fix
- proptest changes in the values we use and the addition of a second case.
- workaround for Osmosis overflow with big numbers of decimals too, which has been mainly replacing checked_sub()s with saturating_sub() to return 0 if it would result in a negative value
- test tube test for overflow

## 4. Checklist

- [x] Does the Readme need to be updated?

## 5. Limitations (optional)

On the `proptest.rs` I was trying to increase the `ACCOUNTS_INITIAL_BALANCE: u128 = 100_000_000_000_000_000;` to something higher, but I was getting liquidityDelta error.

## 6. Future Work (optional)

Enhance further our test suite.

### CODE REVIEW

File: smart-contracts/contracts/cl-vault/src/instantiate.rs
Comment (magiodev) [inline smart-contracts/contracts/cl-vault/src/instantiate.rs:1]: it should be .into() ? i need it to compile

File: smart-contracts/contracts/cl-vault/src/test_tube/proptest.rs [hunk 1]
```diff
@@ -466,7 +473,7 @@ mod tests {
     // get_percentage generates a list of random percentages used to calculate deposit_amount,
     // withdraw_amount, and newers lower and upper ticks based on the previous values
     prop_compose! {
-        fn get_percentage_list()(list in prop::collection::vec(1.0..100.0, ITERATIONS_NUMBER..ITERATIONS_NUMBER+1)) -> Vec<f64> {
+        fn get_percentage_list()(list in prop::collection::vec(1.0..10.0, ITERATIONS_NUMBER..ITERATIONS_NUMBER+1)) -> Vec<f64> {
```
Comment (njerschow) [inline smart-contracts/contracts/cl-vault/src/test_tube/proptest.rs:476]: @magiodev  why was this set to 100? I believe my tests were failing due to insufficient funds because of this.

Comment (magiodev) [reply smart-contracts/contracts/cl-vault/src/test_tube/proptest.rs:476]: this percentage list array is used by every action. the test design is to first take the ITERATIONS_NUMBER, with this value creating a list of random percentages between 1 and 100 and use it for the random action we pick during an iteration.
effectively it could lead to errors as if during an iteration we deposit the 100 percent of a user funds and we try it again later, it should be capped to 99, but it worked till now.

File: smart-contracts/contracts/cl-vault/src/contract.rs [hunk 2]
```diff
@@ -174,9 +179,62 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
...
 }
 
-#[entry_point]
-pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
-    Ok(Response::new().add_attribute("migrate", "successful"))
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    let vault_denom = VAULT_DENOM.load(deps.storage)?;
+
+    let mut total_old = Uint256::zero();
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/contract.rs:186]: As far as I can see, `total_old` is never used

File: smart-contracts/contracts/cl-vault/src/contract.rs [hunk 3]
```diff
@@ -174,9 +179,62 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
...
+            // Convert back to Uint128 for new shares, handling potential overflow
+            let new_shares: Uint128 = Uint128::try_from(new_shares_256)
+                .expect("Conversion from Uint256 to Uint128 failed due to overflow");
+
+            total_old = total_old.checked_add(shares_256)?;
+            Ok((user, new_shares))
+        })
+        .collect();
+
+    for (user, new_shares) in vals? {
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/contract.rs:212]: Why do we split up the for loops into 2 and recalculate certain values?

File: smart-contracts/contracts/cl-vault/src/contract.rs [hunk 4]
```diff
@@ -174,9 +179,62 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
...
+        SHARES.save(deps.storage, user, &new_shares)?;
+
+        // Calculate the burn amount for each user
+        let old_shares_256 = Uint256::from(new_shares.u128() * 10u128.pow(18));
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/contract.rs:216]: This looks very errorprone with rounding to me

File: smart-contracts/contracts/cl-vault/src/contract.rs [hunk 5]
```diff
@@ -174,9 +179,62 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
...
+            let new_shares: Uint128 = Uint128::try_from(new_shares_256)
+                .expect("Conversion from Uint256 to Uint128 failed due to overflow");
+
+            total_old = total_old.checked_add(shares_256)?;
+            Ok((user, new_shares))
+        })
+        .collect();
+
+    for (user, new_shares) in vals? {
+        SHARES.save(deps.storage, user, &new_shares)?;
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/contract.rs:213]: If this save is causing borrow checker issues on deps as mut/immutable, I would prefer to do the save in a seperate loop so we dont have to recalculate old_shares_256

File: smart-contracts/contracts/cl-vault/src/helpers.rs [hunk 6]
```diff
@@ -343,15 +343,17 @@ pub fn get_liquidity_amount_for_unused_funds(
     // then figure out based on current unused balance, what the max initial deposit could be
     // (with the ratio, what is the max tokens we can deposit)
     let tokens = get_unused_balances(deps.storage, &deps.querier, env)?;
+
+    // Notice: checked_sub has been replaced with saturating_sub due to overflowing response from dex
     let unused_t0: Uint256 = tokens
         .find_coin(token0.denom)
         .amount
-        .checked_sub(additional_excluded_funds.0)?
+        .saturating_sub(additional_excluded_funds.0)
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/helpers.rs:351]: Let's revert this change, and instead just reject overflow actions from Osmosis

File: smart-contracts/contracts/cl-vault/src/vault/deposit.rs [hunk 7]
```diff
@@ -129,13 +130,14 @@ pub fn handle_deposit_create_position_reply(
         .parse::<u128>()?
         .into();
 
+    // Notice: checked_sub has been replaced with saturating_sub due to overflowing response from dex
     let refunded = (
-        current_deposit.token0_in.checked_sub(Uint128::new(
+        current_deposit.token0_in.saturating_sub(Uint128::new(
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/vault/deposit.rs:135]: Let's rever this saturating_sub too

File: smart-contracts/contracts/cl-vault/src/vault/deposit.rs [hunk 8]
```diff
@@ -247,13 +249,13 @@ fn refund_bank_msg(
     denom0: String,
     denom1: String,
 ) -> Result<Option<(BankMsg, Vec<Attribute>)>, ContractError> {
+    // Notice: checked_sub has been replaced with saturating_sub due to overflowing response from dex
     let refund0 = current_deposit
         .token0_in
-        .checked_sub(Uint128::new(resp.amount0.parse::<u128>()?))?;
-
+        .saturating_sub(Uint128::new(resp.amount0.parse::<u128>()?));
```
Comment (LaurensKubat) [inline smart-contracts/contracts/cl-vault/src/vault/deposit.rs:255]: Same for this saturating_sub

**Review by LaurensKubat — CHANGES_REQUESTED** [review]
Idea looks good, the migrate message can use some small improvements, Ideally we also have a unit test for the logic in the migrate message

File: smart-contracts/contracts/cl-vault/src/vault/deposit.rs (hunk 8 above)
Comment (ajansari95) [reply smart-contracts/contracts/cl-vault/src/vault/deposit.rs]: [`83cbd87`](https://github.com/quasar-finance/quasar/pull/537/commits/83cbd87458b0beea209d65fdce91cf0511a712cf)

**Review by LaurensKubat — APPROVED** [review]
Lgtm, small nitpick, we left the comments about the saturating sub while leaving the checked sub. In the long term these things cause unnecessary confusion

**Review by ajansari95 — DISMISSED** [review]

**Review by magiodev — APPROVED** [review]
lgtm!

**Review by ajansari95 — DISMISSED** [review]
lgtm

**Review by LaurensKubat — APPROVED** [review]
lgtm

**Comment by LaurensKubat** [comment]
CI is known to be not passing now due to race conditions in CI, fixed in #547
